README.md
memtool
=======

memtool is a program that allows to access memory mapped registers. This is
useful to inspect and modify registers from the command line.

By default memtool uses `/dev/mem` to operate. So to actually work you need to
have `DEVMEM` enabled in the kernel. Note that depending on further kernel
configuration you can only access unused io-memory, see the kernel
configuration knobs `IO_STRICT_DEVMEM` and `STRICT_DEVMEM`. Also note that
there might be further restrictions. E.g. on several i.MX ARM SoCs register
access is not possible from user mode for certain areas unless configured
otherwise.

memtool can also operate on plain files, and access PHY registers (via the
`ioctl`s `SIOCSMIIREG` and `SIOCGMIIREG`).

Examples
---------

 * Write a 32-bit wide 0 to register at address 0x73f00040

    ```sh
    # memtool mw -l 0x73f00040 0
    ```

 * Read the first two registers of the default PHY attached to network device
   `eth0`:

    ```sh
    # memtool md -s mdio:eth0. -w 0+4
    ```

 * Place a (usually red) dot on the framebuffer:

    ```sh
    # memtool mw -d /dev/fb0 -w 0 0xfc00
    ```



debian/changelog
memtool (2018.03.0-1) unstable; urgency=medium

  * Switch to Priority: optional.
  * Use https in the copyright-format URL.
  * Standards-Version: 4.1.2 (no further changes needed)
  * New upstream release
    - Support to access network phys
  * Build using debhelper 11

 -- Uwe Kleine-König <ukleinek@debian.org>  Fri, 16 Mar 2018 12:19:04 +0100

memtool (2016.10.0-1) unstable; urgency=medium

  * New upstream release fixing handling of small files (Closes: #831219).
  * Standards-Version: 3.9.8 (no changes needed)

 -- Uwe Kleine-König <ukleinek@debian.org>  Fri, 21 Oct 2016 21:16:03 +0200

memtool (2016.07.0-1) unstable; urgency=medium

  * New upstream release fixing memtool md on read-only files.

 -- Uwe Kleine-König <ukleinek@debian.org>  Thu, 14 Jul 2016 09:27:32 +0200

memtool (2016.01.0-2) unstable; urgency=medium

  Make lintian happy:
  * Fix order of paragraphs in d/copyright to please lintian (and
    correctness).
  * Check signatures in d/watch to please lintian even more.
  * Enable all hardening features.

 -- Uwe Kleine-König <ukleinek@debian.org>  Mon, 18 Apr 2016 21:50:27 +0200

memtool (2016.01.0-1) unstable; urgency=medium

  * New upstream release which fixes LFS issues.
  * Build also on non-linux platforms.

 -- Uwe Kleine-König <ukleinek@debian.org>  Wed, 06 Jan 2016 00:04:33 +0100

memtool (2015.12.2-1) unstable; urgency=medium

  * New upstream release which among others provides a manpage now.
  * Add Vcs-{Browser,Git} to d/control.

 -- Uwe Kleine-König <ukleinek@debian.org>  Fri, 11 Dec 2015 11:05:46 +0100

memtool (2015.12.0-1) unstable; urgency=medium

  * Initial release for Debian (Closes: #807315).

 -- Uwe Kleine-König <ukleinek@debian.org>  Mon, 07 Dec 2015 20:45:44 +0100


acc_mdio.c
/*
 * Copyright (C) 2018 Pengutronix, Uwe Kleine-König <oss-tools@pengutronix.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <unistd.h>

#include <linux/if.h>
#include <linux/mii.h>
#include <linux/sockios.h>

#include "fileaccpriv.h"

#define container_of(ptr, type, member) \
	(type *)((char *)(ptr) - (char *) &((type *)0)->member)

struct memtool_mdio_fd {
	struct memtool_fd mfd;
	int fd;
	uint16_t phy_id;
	char ifrn_name[IFNAMSIZ];
};

static ssize_t mdio_read(struct memtool_fd *handle, off_t offset,
			 void *buf, size_t nbytes, int width)
{
	struct memtool_mdio_fd *mdio_fd =
		container_of(handle, struct memtool_mdio_fd, mfd);
	struct ifreq ifr;
	struct mii_ioctl_data *mii = (void *)&ifr.ifr_data;
	size_t i = 0;
	int ret;

	if (width != 2) {
		fprintf(stderr,
			"mdio can only be accessed with memory width 2\n");
		return -EINVAL;
	}

	assert((nbytes & 1) == 0);

	memset(&ifr, 0, sizeof(ifr));
	strcpy(ifr.ifr_name, mdio_fd->ifrn_name);

	mii->phy_id = mdio_fd->phy_id;

	while (2 * i < nbytes) {
		mii->reg_num = offset / 2 + i;

		ret = ioctl(mdio_fd->fd, SIOCGMIIREG, &ifr);
		if (ret < 0) {
			perror("Failure to read register");
			return -1;
		}

		((uint16_t *)buf)[i] = mii->val_out;
		++i;
	}

	return 2 * i;
}

static ssize_t mdio_write(struct memtool_fd *handle, off_t offset,
			  const void *buf, size_t nbytes, int width)
{
	struct memtool_mdio_fd *mdio_fd =
		container_of(handle, struct memtool_mdio_fd, mfd);
	struct ifreq ifr;
	struct mii_ioctl_data *mii = (void *)&ifr.ifr_data;
	size_t i = 0;
	int ret;

	if (width != 2) {
		fprintf(stderr,
			"mdio can only be accessed with memory width 2\n");
		return -EINVAL;
	}

	assert((nbytes & 1) == 0);

	memset(&ifr, 0, sizeof(ifr));
	strcpy(ifr.ifr_name, mdio_fd->ifrn_name);

	mii->phy_id = mdio_fd->phy_id;

	while (2 * i < nbytes) {
		mii->reg_num = offset / 2 + i;
		mii->val_in = ((uint16_t *)buf)[i];

		ret = ioctl(mdio_fd->fd, SIOCSMIIREG, &ifr);
		if (ret < 0) {
			perror("Failure to write register");
			return -1;
		}

		++i;
	}

	return 2 * i;
}

static int mdio_close(struct memtool_fd *handle)
{
	struct memtool_mdio_fd *mdio_fd =
		container_of(handle, struct memtool_mdio_fd, mfd);
	int ret;

	ret = close(mdio_fd->fd);

	free(mdio_fd);

	return ret;
}

struct memtool_fd *mdio_open(const char *spec, int flags)
{
	struct memtool_mdio_fd *mdio_fd;
	char *delim;
	char *endp;
	long int val;

	mdio_fd = malloc(sizeof(*mdio_fd));
	if (!mdio_fd) {
		fprintf(stderr, "Failure to allocate mdio_fd\n");
		return NULL;
	}

	mdio_fd->mfd.read = mdio_read;
	mdio_fd->mfd.write = mdio_write;
	mdio_fd->mfd.close = mdio_close;

	mdio_fd->fd = socket(AF_INET, SOCK_DGRAM, 0);
	if (mdio_fd->fd < 0) {
		perror("socket");
		goto err_socket;
	}

	delim = strrchr(spec, '.');
	if (!delim) {
		fprintf(stderr, "Failed to parse phy specifier, no \".\"\n");
		goto err_parse;
	}

	if (delim - spec >= sizeof(mdio_fd->ifrn_name)) {
		fprintf(stderr, "device string too long\n");
		goto err_parse;
	}

	memcpy(mdio_fd->ifrn_name, spec, delim - spec);
	mdio_fd->ifrn_name[delim - spec] = '\0';

	val = strtol(delim + 1, &endp, 0);
	if (*endp != '\0') {
		fprintf(stderr, "failed to parse phy_id\n");
		goto err_parse;
	}

	if (val < 0 || val >= (1 << 16)) {
		fprintf(stderr, "phy_id out of range\n");
err_parse:
		close(mdio_fd->fd);
err_socket:
		free(mdio_fd);
		return NULL;
	}
	mdio_fd->phy_id = val;

	return &mdio_fd->mfd;
}


acc_mmap.c
/*
 * Copyright (C) 2018 Pengutronix, Uwe Kleine-König <oss-tools@pengutronix.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <unistd.h>

#include "fileaccpriv.h"

#define container_of(ptr, type, member) \
	(type *)((char *)(ptr) - (char *) &((type *)0)->member)

static off_t mmap_pagesize(void) __attribute__((const));
static off_t mmap_pagesize(void)
{
	static off_t pagesize;

	if (pagesize == 0)
		pagesize = sysconf(_SC_PAGE_SIZE);

	if (pagesize == 0)
		pagesize = 4096;

	return pagesize;
}

struct memtool_mmap_fd {
	struct memtool_fd mfd;
	struct stat s;
	int fd;
};

static ssize_t mmap_read(struct memtool_fd *handle, off_t offset,
			 void *buf, size_t nbytes, int width)
{
	struct memtool_mmap_fd *mmap_fd =
		container_of(handle, struct memtool_mmap_fd, mfd);
	struct stat *s = &mmap_fd->s;
	off_t map_start, map_off;
	void *map;
	size_t i = 0;
	int ret;

	if (S_ISREG(s->st_mode)) {
		if (s->st_size <= offset) {
			errno = EINVAL;
			perror("File to small");
			return -1;
		}

		if (s->st_size < offset + nbytes)
			/* truncating */
			nbytes = s->st_size - offset;
	}

	map_start = offset & ~(mmap_pagesize() - 1);
	map_off = offset - map_start;

	map = mmap(NULL, nbytes + map_off, PROT_READ,
		   MAP_SHARED, mmap_fd->fd, map_start);
	if (map == MAP_FAILED) {
		perror("mmap");
		return -1;
	}

	while (i * width + width <= nbytes) {
		switch (width) {
		case 1:
			((uint8_t *)buf)[i] = ((uint8_t *)(map + map_off))[i];
			break;
		case 2:
			((uint16_t *)buf)[i] = ((uint16_t *)(map + map_off))[i];
			break;
		case 4:
			((uint32_t *)buf)[i] = ((uint32_t *)(map + map_off))[i];
			break;
		case 8:
			((uint64_t *)buf)[i] = ((uint64_t *)(map + map_off))[i];
			break;
		}
		++i;
	}

	ret = munmap(map, nbytes + map_off);
	if (ret < 0) {
		perror("munmap");
		return -1;
	}

	return i * width;
}

static ssize_t mmap_write(struct memtool_fd *handle, off_t offset,
			  const void *buf, size_t nbytes, int width)
{
	struct memtool_mmap_fd *mmap_fd =
		container_of(handle, struct memtool_mmap_fd, mfd);
	struct stat *s = &mmap_fd->s;
	off_t map_start, map_off;
	void *map;
	size_t i = 0;
	int ret;

	if (S_ISREG(s->st_mode) && s->st_size < offset + nbytes) {
		ret = posix_fallocate(mmap_fd->fd, offset, nbytes);
		if (ret) {
			errno = ret;
			perror("fallocate");
			return -1;
		}
		s->st_size = offset + nbytes;
	}

	map_start = offset & ~(mmap_pagesize() - 1);
	map_off = offset - map_start;

	map = mmap(NULL, nbytes + map_off, PROT_WRITE,
		   MAP_SHARED, mmap_fd->fd, map_start);
	if (map == MAP_FAILED) {
		perror("mmap");
		return -1;
	}

	while (i * width + width <= nbytes) {
		switch (width) {
		case 1:
			((uint8_t *)(map + map_off))[i] = ((uint8_t *)buf)[i];
			break;
		case 2:
			((uint16_t *)(map + map_off))[i] = ((uint16_t *)buf)[i];
			break;
		case 4:
			((uint32_t *)(map + map_off))[i] = ((uint32_t *)buf)[i];
			break;
		case 8:
			((uint64_t *)(map + map_off))[i] = ((uint64_t *)buf)[i];
			break;
		}
		++i;
	}

	ret = munmap(map, nbytes + map_off);
	if (ret < 0) {
		perror("munmap");
		return -1;
	}

	return i * width;
}

static int mmap_close(struct memtool_fd *handle)
{
	struct memtool_mmap_fd *mmap_fd =
		container_of(handle, struct memtool_mmap_fd, mfd);
	int ret;

	ret = close(mmap_fd->fd);

	free(mmap_fd);

	return ret;
}

struct memtool_fd *mmap_open(const char *spec, int flags)
{
	struct memtool_mmap_fd *mmap_fd;
	int ret;

	mmap_fd = malloc(sizeof(*mmap_fd));
	if (!mmap_fd) {
		fprintf(stderr, "Failure to allocate mmap_fd\n");
		return NULL;
	}

	mmap_fd->mfd.read = mmap_read;
	mmap_fd->mfd.write = mmap_write;
	mmap_fd->mfd.close = mmap_close;

	mmap_fd->fd = open(spec, flags, S_IRUSR | S_IWUSR);
	if (mmap_fd->fd < 0) {
		perror("open");
		free(mmap_fd);
		return NULL;
	}

	ret = fstat(mmap_fd->fd, &mmap_fd->s);
	if (ret) {
		perror("fstat");
		close(mmap_fd->fd);
		free(mmap_fd);
		return NULL;
	}

	return &mmap_fd->mfd;
}


fileaccess.c
/*
 * Copyright (C) 2018 Pengutronix, Uwe Kleine-König <oss-tools@pengutronix.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <string.h>
#include <stdio.h>

#include "fileaccess.h"
#include "fileaccpriv.h"

void *memtool_open(const char *spec, int flags)
{
	if (!strncmp(spec, "mmap:", 5)) {
		return mmap_open(spec + 5, flags);
	} else if (!strncmp(spec, "mdio:", 5)) {
#ifdef USE_MDIO
		return mdio_open(spec + 5, flags);
#else
		fprintf(stderr, "mdio support not compiled in\n");
		return NULL;
#endif
	} else {
		return mmap_open(spec, flags);
	}
}

ssize_t memtool_read(void *handle,
		     off_t offset, void *buf, size_t nbytes, int width)
{
	struct memtool_fd *mfd = handle;

	return mfd->read(mfd, offset, buf, nbytes, width);
}

ssize_t memtool_write(void *handle,
		      off_t offset, const void *buf, size_t nbytes, int width)
{
	struct memtool_fd *mfd = handle;

	return mfd->write(mfd, offset, buf, nbytes, width);
}

int memtool_close(void *handle)
{
	struct memtool_fd *mfd = handle;

	return mfd->close(mfd);
}


fileaccess.h
/*
 * Copyright (C) 2018 Pengutronix, Uwe Kleine-König <oss-tools@pengutronix.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <sys/types.h>

void *memtool_open(const char *spec, int flags);
ssize_t memtool_read(void *handle, off_t offset,
		     void *buf, size_t nbytes, int width);
ssize_t memtool_write(void *handle, off_t offset,
		      const void *buf, size_t nbytes, int width);
int memtool_close(void *handle);


fileaccpriv.h
/*
 * Copyright (C) 2018 Pengutronix, Uwe Kleine-König <oss-tools@pengutronix.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
#include <sys/types.h>

struct memtool_fd {
	ssize_t (*read)(struct memtool_fd *handle, off_t offset,
			void *buf, size_t nbytes, int width);
	ssize_t (*write)(struct memtool_fd *handle, off_t offset,
			 const void *buf, size_t nbytes, int width);
	int (*close)(struct memtool_fd *handle);
};

struct memtool_fd *mdio_open(const char *spec, int flags);
struct memtool_fd *mmap_open(const char *spec, int flags);


memtool.c
/*
 * Copyright (C) 2015 Pengutronix, Sascha Hauer <entwicklung@pengutronix.de>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 */

#include <assert.h>
#include <libgen.h>
#include <stdint.h>
#include <stdio.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <getopt.h>
#include <unistd.h>
#include <errno.h>
#include <stdlib.h>
#include <stdint.h>
#include <ctype.h>
#include <string.h>
#include <inttypes.h>

#include "fileaccess.h"

#define DISP_LINE_LEN	16

/*
 * Like strtoull() but handles an optional G, M, K or k
 * suffix for Gibibyte, Mibibyte or Kibibyte.
 */
static unsigned long long strtoull_suffix(const char *str, char **endp, int base)
{
	unsigned long long val;
	char *end;

	val = strtoull(str, &end, base);

	switch (*end) {
	case 'G':
		val *= 1024;
	case 'M':
		val *= 1024;
	case 'k':
	case 'K':
		val *= 1024;
		end++;
	default:
		break;
	}

	if (endp)
		*endp = (char *)end;

	return val;
}

/*
 * This function parses strings in the form <startadr>[-endaddr]
 * or <startadr>[+size] and fills in start and size accordingly.
 * <startadr> and <endadr> can be given in decimal or hex (with 0x prefix)
 * and can have an optional G, M, K or k suffix.
 *
 * examples:
 * 0x1000-0x2000 -> start = 0x1000, size = 0x1001
 * 0x1000+0x1000 -> start = 0x1000, size = 0x1000
 * 0x1000        -> start = 0x1000, size = ~0
 * 1M+1k         -> start = 0x100000, size = 0x400
 */
static int parse_area_spec(const char *str, off_t *start, size_t *size)
{
	char *endp;
	off_t end;

	if (!isdigit(*str))
		return -1;

	*start = strtoull_suffix(str, &endp, 0);

	str = endp;

	if (!*str) {
		/* beginning given, but no size, assume maximum size */
		*size = ~0;
		return 0;
	}

	if (*str == '-') {
		/* beginning and end given */
		end = strtoull_suffix(str + 1, NULL, 0);
		if (end < *start) {
			fprintf(stderr, "end < start\n");
			return -1;
		}
		*size = end - *start + 1;
		return 0;
	}

	if (*str == '+') {
		/* beginning and size given */
		*size = strtoull_suffix(str + 1, NULL, 0);
		return 0;
	}

	return -1;
}
#define swab64(x) ((uint64_t)(						\
	(((uint64_t)(x) & (uint64_t)0x00000000000000ffUL) << 56) |	\
	(((uint64_t)(x) & (uint64_t)0x000000000000ff00UL) << 40) |	\
	(((uint64_t)(x) & (uint64_t)0x0000000000ff0000UL) << 24) |	\
	(((uint64_t)(x) & (uint64_t)0x00000000ff000000UL) <<  8) |	\
	(((uint64_t)(x) & (uint64_t)0x000000ff00000000UL) >>  8) |	\
	(((uint64_t)(x) & (uint64_t)0x0000ff0000000000UL) >> 24) |	\
	(((uint64_t)(x) & (uint64_t)0x00ff000000000000UL) >> 40) |	\
	(((uint64_t)(x) & (uint64_t)0xff00000000000000UL) >> 56)))

#define swab32(x) ((uint32_t)(						\
	(((uint32_t)(x) & (uint32_t)0x000000ffUL) << 24) |		\
	(((uint32_t)(x) & (uint32_t)0x0000ff00UL) <<  8) |		\
	(((uint32_t)(x) & (uint32_t)0x00ff0000UL) >>  8) |		\
	(((uint32_t)(x) & (uint32_t)0xff000000UL) >> 24)))

#define swab16(x) ((uint16_t)(						\
	(((uint16_t)(x) & (uint16_t)0x00ffU) << 8) |			\
	(((uint16_t)(x) & (uint16_t)0xff00U) >> 8)))

static int memory_display(const void *addr, off_t offs,
			  size_t nbytes, int width, int swab)
{
	size_t linebytes, i;
	u_char	*cp;

	/* Print the lines.
	 *
	 * We buffer all read data, so we can make sure data is read only
	 * once, and all accesses are with the specified bus width.
	 */
	do {
		char linebuf[DISP_LINE_LEN];
		uint64_t *uqp = (uint64_t *)linebuf;
		uint32_t *uip = (uint32_t *)linebuf;
		uint16_t *usp = (uint16_t *)linebuf;
		uint8_t *ucp = (uint8_t *)linebuf;
		unsigned count = 52;

		printf("%08llx:", (unsigned long long)offs);
		linebytes = (nbytes > DISP_LINE_LEN) ? DISP_LINE_LEN : nbytes;

		for (i = 0; i < linebytes; i += width) {
			if (width == 8) {
				uint64_t res;
				res = (*uqp++ = *((uint64_t *)addr));
				if (swab)
					res = swab64(res);
				count -= printf(" %016" PRIx64, res);
			} else if (width == 4) {
				uint32_t res;
				res = (*uip++ = *((uint *)addr));
				if (swab)
					res = swab32(res);
				count -= printf(" %08" PRIx32, res);
			} else if (width == 2) {
				uint16_t res;
				res = (*usp++ = *((ushort *)addr));
				if (swab)
					res = swab16(res);
				count -= printf(" %04" PRIx16, res);
			} else {
				count -= printf(" %02x", (*ucp++ = *((u_char *)addr)));
			}
			addr += width;
			offs += width;
		}

		while (count--)
			putchar(' ');

		cp = (uint8_t *)linebuf;
		for (i = 0; i < linebytes; i++) {
			if ((*cp < 0x20) || (*cp > 0x7e))
				putchar('.');
			else
				printf("%c", *cp);
			cp++;
		}

		putchar('\n');
		nbytes -= linebytes;
	} while (nbytes > 0);

	return 0;
}

static void usage_md(void)
{
	printf(
"md - memory display\n"
"\n"
"Usage: md [-bwlqsx] REGION\n"
"\n"
"Display (hex dump) a memory region.\n"
"\n"
"Options:\n"
"  -b        byte access\n"
"  -w        word access (16 bit)\n"
"  -l        long access (32 bit)\n"
"  -q        quad access (64 bit)\n"
"  -s <FILE> display file (default /dev/mem)\n"
"  -x        swap bytes at output\n"
"\n"
"Memory regions can be specified in two different forms: START+SIZE\n"
"or START-END, If START is omitted it defaults to 0x100\n"
"Sizes can be specified as decimal, or if prefixed with 0x as hexadecimal.\n"
"An optional suffix of k, M or G is for kbytes, Megabytes or Gigabytes.\n"
	);

}

static int cmd_memory_display(int argc, char **argv)
{
	int opt;
	int width = 4;
	size_t bufsize, size = 0x100;
	char *buf;
	void *handle;
	off_t start = 0x0;
	char *file = "/dev/mem";
	int swap = 0;

	while ((opt = getopt(argc, argv, "bwlqs:xh")) != -1) {
		switch (opt) {
		case 'b':
			width = 1;
			break;
		case 'w':
			width = 2;
			break;
		case 'l':
			width = 4;
			break;
		case 'q':
			width = 8;
			break;
		case 's':
			file = optarg;
			break;
		case 'x':
			swap = 1;
			break;
		case 'h':
			usage_md();
			return 0;
		}
	}

	if (optind < argc) {
		if (parse_area_spec(argv[optind], &start, &size)) {
			fprintf(stderr, "could not parse: %s\n", argv[optind]);
			return EXIT_FAILURE;
		}
		if (size == ~0)
			size = 0x100;
	}

	if (size & (width - 1)) {
		size &= ~(width - 1);
		fprintf(stderr, "warning: skipping truncated read, size=%zu\n",
			size);
	}

	if (!size)
		return EXIT_SUCCESS;

	bufsize = size;
	if (bufsize > 4096)
		bufsize = 4096;

	buf = malloc(bufsize);
	if (!buf) {
		fprintf(stderr, "could not allocate memory\n");
		return EXIT_FAILURE;
	}

	handle = memtool_open(file, O_RDONLY);
	if (!handle)
		return EXIT_FAILURE;

	while (size) {
		int ret;

		if (size < bufsize)
			bufsize = size;

		ret = memtool_read(handle, start, buf, bufsize, width);
		if (ret < 0)
			return EXIT_FAILURE;

		assert(ret == bufsize);
		memory_display(buf, start, bufsize, width, swap);

		start += bufsize;
		size -= bufsize;
	}

	memtool_close(handle);

	return EXIT_SUCCESS;
}

static void usage_mw(void)
{
	printf(
"mw - memory write\n"
"\n"
"Usage: mw [-bwlqd] OFFSET DATA...\n"
"\n"
"Write DATA value(s) to the specified REGION.\n"
"\n"
"Options:\n"
"  -b        byte access\n"
"  -w        word access (16 bit)\n"
"  -l        long access (32 bit)\n"
"  -q        quad access (64 bit)\n"
"  -d <FILE> write file (default /dev/mem)\n"
	);
}

static int cmd_memory_write(int argc, char *argv[])
{
	off_t adr;
	size_t bufsize, size;
	char *buf;
	void *handle;
	int width = 4;
	int opt;
	int i, ret;
	char *file = "/dev/mem";

	while ((opt = getopt(argc, argv, "bwlqd:h")) != -1) {
		switch (opt) {
		case 'b':
			width = 1;
			break;
		case 'w':
			width = 2;
			break;
		case 'l':
			width = 4;
			break;
		case 'q':
			width = 8;
			break;
		case 'd':
			file = optarg;
			break;
		case 'h':
			usage_mw();
			return 0;
		}
	}

	if (optind + 1 >= argc) {
		fprintf(stderr, "Too few parameters for mw\n");
		return EXIT_FAILURE;
	}

	adr = strtoull_suffix(argv[optind++], NULL, 0);

	size = (argc - optind) * width;
	if (!size)
		return EXIT_SUCCESS;

	bufsize = size;
	if (bufsize > 4096)
		bufsize = 4096;

	buf = malloc(bufsize);
	if (!buf) {
		fprintf(stderr, "could not allocate memory\n");
		return EXIT_FAILURE;
	}

	handle = memtool_open(file, O_RDWR | O_CREAT);
	if (!handle)
		return EXIT_FAILURE;

	while (optind < argc) {
		i = 0;

		while (optind < argc && i * width < bufsize) {
			switch (width) {
			case 1:
				((uint8_t *)buf)[i] =
					strtoull(argv[optind], NULL, 0);
				break;
			case 2:
				((uint16_t *)buf)[i] =
					strtoull(argv[optind], NULL, 0);
				break;
			case 4:
				((uint32_t *)buf)[i] =
					strtoull(argv[optind], NULL, 0);
				break;
			case 8:
				((uint64_t *)buf)[i] =
					strtoull(argv[optind], NULL, 0);
				break;
			}
			++i;
			++optind;
		}

		ret = memtool_write(handle, adr, buf, i * width, width);
		if (ret < 0)
			break;

		assert(ret == i * width);
		adr += i * width;
	}


	memtool_close(handle);
	free(buf);

	return ret < 0 ? EXIT_FAILURE : EXIT_SUCCESS;
}

struct cmd {
	int (*cmd)(int argc, char **argv);
	const char *name;
};

#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))

static struct cmd cmds[] = {
	{
		.cmd = cmd_memory_display,
		.name = "md",
	}, {
		.cmd = cmd_memory_write,
		.name = "mw",
	},
};

static void usage(void)
{
	printf(
"memtool - display and modify memory\n"
"\n"
"Usage: memtool <cmd> [OPTIONS]\n"
"\n"
"memtool is divided into subcommands. Supported commands are:\n"
"md: memory display, Show regions of memory\n"
"mw: memory write, write values to memory\n"
"\n"
"To show help for a subcommand do 'memtool <cmd> -h'\n"
"\n"
"memtool is a collection of tools to show (hexdump) and modify arbitrary files.\n"
"By default /dev/mem is used to allow access to physical memory.\n"
	);
}

int main(int argc, char **argv)
{
	int i;
	struct cmd *cmd;

	if (!strcmp(basename(argv[0]), "memtool")) {
		argv++;
		argc--;

		if (argc > 0 && !strcmp(argv[0], "-V")) {
			printf("%s\n", PACKAGE_STRING);
			return EXIT_SUCCESS;
		}
	}

	if (argc < 1) {
		fprintf(stderr, "No command given\n");
		usage();
		return EXIT_FAILURE;
	}

	for (i = 0; i < ARRAY_SIZE(cmds); i++) {
		cmd = &cmds[i];
		if (!strcmp(argv[0], cmd->name))
			return cmd->cmd(argc, argv);
	}

	fprintf(stderr, "No such command: %s\n", argv[0]);

	usage();

	return EXIT_FAILURE;
}
